torch/_library/autograd.py

```diff
@@ -31,8 +32,95 @@ def backward(ctx, *grads):
         },
     )
 
+    schema = opdef._opoverload._schema
+    if any(
+        utils.is_tensorlist_like_type(a.type)
+        for a in (*schema.arguments, *schema.returns)
+    ):
+        Generated = supports_tensorlist(Generated)
+
     def autograd_impl(*args):
         result = Generated.apply(*args)  # type: ignore[attr-defined]
         return result
 
     return autograd_impl
+
+
+def supports_tensorlist(cls: Any) -> Any:
```

Comment:
There is quite a bit of state here.
I think it would be good to have a few asserts to make sure that we only ever do apply -> fw -> bw -> apply -> fw -> bw -> etc
And that any double apply/bw/etc will raise a nice error saying this is not supported yet.

Reply:
That's a good catch. Let me think about if we can support the double apply situation -- that is probably going to be common.

Reply:
Updated with nice error messages. I also expand the number of things we support: we can support double apply. The secret is to put the input_spec/output_spec onto the ctx object so that each of the backward have their own input_spec/output_spec